awsproviderlint: New Check: Prefer aws_ami data source instead of hardcoded Amazon Machine Image identifiers

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Description

<!--- Please leave a helpful description of the feature request here. --->

To ensure test configurations are region and partition agnostic, any hardcoded AMI identifiers (ami-[a-f0-9]{8,17}) should be replaced with the `aws_ami` data source.

**This check was added** in #14031 and is [AWSAT002](https://github.com/terraform-providers/terraform-provider-aws/tree/master/awsproviderlint/passes/AWSAT002). We'll keep this issue open while fixes are made that are identified by the check.

#### Flagged Code

In test configuration:
```
ami-123456
```
#### Passing Code

In test configuration:
```
data "aws_ami" "example" {
  # ... configuration ...
}
```

### To Do

- [x] data_source_aws_instance_test.go (#14145)
- [x] data_source_aws_launch_configuration_test.go (#14147)
- [x] resource_aws_autoscaling_attachment_test.go (#14148)
- [x] resource_aws_autoscaling_group_test.go (#14222)
- [x] resource_aws_autoscaling_lifecycle_hook_test.go (#14321)
- [x] resource_aws_autoscaling_notification_test.go (#14323)
- [x] resource_aws_autoscaling_schedule_test.go (#14325)
- [x] resource_aws_eip_association_test.go (#14232)
- [x] resource_aws_elb_attachment_test.go (#14235)
- [ ] resource_aws_instance_test.go (#14159, original fix) (#15143, second fix)
- [x] resource_aws_launch_template_test.go (#14233)
- [x] resource_aws_network_interface_attachment_test.go (#14234)
- [x] resource_aws_network_interface_test.go (#14330)
- [x] resource_aws_route_test.go (#14174)
- [x] resource_aws_spot_fleet_request_test.go (#14083)
- [x] resource_aws_spot_instance_request_test.go (#14082)
- [x] validators_test.go (#15142)

@YakDriver this looks really close to complete, do you think we could get this turned on this week?

```console
$ awsproviderlint -AWSAT002 ./aws
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_instance_test.go:3836:46: AWSAT002: AMI IDs should not be hardcoded
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_instance_test.go:4308:53: AWSAT002: AMI IDs should not be hardcoded
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/validators_test.go:1182:14: AWSAT002: AMI IDs should not be hardcoded
```

--- a/GNUmakefile
+++ b/GNUmakefile
@@ -99,6 +99,7 @@ awsproviderlint:
 		-AT007 \
 		-AT008 \
 		-AWSAT001 \
+		-AWSAT002 \
 		-AWSAT004 \
 		-AWSR001 \
 		-AWSR002 \

--- a/aws/provider_test.go
+++ b/aws/provider_test.go
@@ -558,13 +558,6 @@ func testAccRegionPreCheck(t *testing.T, region string) {
 	}
 }
 
-// testAccPartitionPreCheck checks that the test partition is the specified partition.
-func testAccPartitionPreCheck(t *testing.T, partition string) {
-	if testAccGetPartition() != partition {
-		t.Skipf("skipping tests; partition (%s) does not equal %s", testAccGetPartition(), partition)
-	}
-}
-
 func testAccOrganizationsAccountPreCheck(t *testing.T) {
 	conn := testAccProvider.Meta().(*AWSClient).organizationsconn
 	input := &organizations.DescribeOrganizationInput{}

--- a/aws/resource_aws_instance_test.go
+++ b/aws/resource_aws_instance_test.go
@@ -922,30 +922,6 @@ func TestAccAWSInstance_ipv6_supportAddressCountWithIpv4(t *testing.T) {
 	})
 }
 
-func TestAccAWSInstance_multipleRegions(t *testing.T) {
-	var v ec2.Instance
-	resourceName := "aws_instance.test"
-
-	// record the initialized providers so that we can use them to
-	// check for the instances in each region
-	var providers []*schema.Provider
-
-	resource.ParallelTest(t, resource.TestCase{
-		PreCheck:          func() { testAccPreCheck(t); testAccPartitionPreCheck(t, "aws") },
-		ProviderFactories: testAccProviderFactories(&providers),
-		CheckDestroy:      testAccCheckWithProviders(testAccCheckInstanceDestroyWithProvider, &providers),
-		Steps: []resource.TestStep{
-			{
-				Config: testAccInstanceConfigMultipleRegions,
-				Check: resource.ComposeTestCheckFunc(
-					testAccCheckInstanceExistsWithProvider(resourceName, &v, testAccAwsRegionProviderFunc("us-west-2", &providers)),
-					testAccCheckInstanceExistsWithProvider("aws_instance.test2", &v, testAccAwsRegionProviderFunc("us-east-1", &providers)),
-				),
-			},
-		},
-	})
-}
-
 func TestAccAWSInstance_NetworkInstanceSecurityGroups(t *testing.T) {
 	var v ec2.Instance
 	resourceName := "aws_instance.test"
@@ -1398,7 +1374,7 @@ func TestAccAWSInstance_rootBlockDeviceMismatch(t *testing.T) {
 	rName := acctest.RandomWithPrefix("tf-acc-test")
 
 	resource.ParallelTest(t, resource.TestCase{
-		PreCheck:     func() { testAccPreCheck(t); testAccRegionPreCheck(t, "us-west-2") },
+		PreCheck:     func() { testAccPreCheck(t); testAccRegionPreCheck(t, "us-west-2") }, //lintignore:AWSAT003
 		Providers:    testAccProviders,
 		CheckDestroy: testAccCheckInstanceDestroy,
 		Steps: []resource.TestStep{
@@ -3833,29 +3809,6 @@ resource "aws_instance" "test" {
 `, rName)
 }
 
-const testAccInstanceConfigMultipleRegions = `
-provider "awswest" {
-  region = "us-west-2"
-}
-
-provider "awseast" {
-  region = "us-east-1"
-}
-resource "aws_instance" "test" {
-  # us-west-2
-  provider      = "awswest"
-  ami           = "ami-4fccb37f"
-  instance_type = "m1.small"
-}
-
-resource "aws_instance" "test2" {
-  # us-east-1
-  provider      = "awseast"
-  ami           = "ami-8c6ea9e4"
-  instance_type = "m1.small"
-}
-`
-
 func testAccCheckInstanceConfigTags() string {
 	return composeConfig(testAccLatestAmazonLinuxHvmEbsAmiConfig(), fmt.Sprintf(`
 resource "aws_instance" "test" {
@@ -4316,7 +4269,7 @@ resource "aws_instance" "test" {
     volume_size = 13
   }
 }
-`
+` //lintignore:AWSAT002
 }
 
 func testAccInstanceConfigForceNewAndTagsDrift(rName string) string {

